main_test.go

```diff
@@ -0,0 +1,48 @@
+package main
+
+import (
+	"testing"
+	"net/http"
+	"time"
+	"github.com/pkg/errors"
+	"github.com/nishanths/fullstory"
+)
+
+func TestGetRetryInfo(t *testing.T) {
+	testCases := []struct {
+		err           error
+		expDoRetry    bool
+		expRetryAfter time.Duration
+	}{
+		{
+			err:           errors.New("random error!"),
+			expDoRetry:    true,
+			expRetryAfter: defaultRetryAfterDuration,
+		},
+		{
+			err:           fullstory.StatusError{StatusCode: http.StatusTooManyRequests, RetryAfter: 3 * time.Second},
+			expDoRetry:    true,
+			expRetryAfter: 3 * time.Second,
+		},
+		{
+			err:           fullstory.StatusError{StatusCode: http.StatusInternalServerError, RetryAfter: 3 * time.Second},
```

Comment:
can you change the status code for this test to StatusServiceUnavailable? It is the 5xx status that actually returns a retry-after. Protects in case someone in the future tries to make the retry logic target more specific 5xx errors.

Reply:
Actually the intent here is to retry on all 5xx errors, and when there is no `Retry-After` header, to use our default duration, regardless of the specific code. The general argument is that you don't know enough about the error (even a 500) to say that retrying is a lost cause, so for the sake of robustness you should give it a go. I've seen cases where some of our internal microservices returned a 500 because of some temporary blip in GCP that we didn't interpret appropriately; we just categorized it as a failure and called it a day.

Reply:
Agreed that we should retry all 5xx errors. Its just that 503 errors are most likely to specify a custom retry-after header, hence the recommendation. The 503 tests 5xx errors get retried and also protects against people in the future erroneously ignoring 503 retry-after directives!